modules/cats/test/src/HashVisitorSpec.scala

```diff
@@ -0,0 +1,266 @@
+import smithy4s.interopcats.testcases.IntOrInt
+import smithy4s.interopcats.testcases.RecursiveFoo
+import weaver.FunSuite
+
+object HashVisitorSpec extends FunSuite {
+
+  val visitor: SchemaVisitor[Hash] = SchemaVisitorHash
+
+  test("int") {
+    val schema: Schema[Int] = int
+    val intValue = 1
+    val hashOutput = visitor(schema).hash(intValue)
+    expect.eql(intValue.hashCode, hashOutput)
+  }
+
+  test("string") {
+    val schema: Schema[String] = string
+    val foo = "foo"
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql("foo".hashCode, hashOutput)
+  }
+
+  test("boolean") {
+    val schema: Schema[Boolean] = boolean
+    val foo = true
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+  test("long") {
+    val schema: Schema[Long] = long
+    val foo = 1L
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+  test("short") {
+    val schema: Schema[Short] = short
+    val foo = 1.toShort
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+  test("byte") {
+    val schema: Schema[Byte] = byte
+    val foo = 1.toByte
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+
+  }
+
+  test("double") {
+    val schema: Schema[Double] = double
+    val foo = 1.0
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+
+  }
+
+  test("float") {
+    val schema: Schema[Float] = float
+    val foo = 1.0f
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+  test("bigint") {
+    val schema: Schema[BigInt] = bigint
+    val foo = BigInt(1)
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+  test("bigdecimal") {
+    val schema: Schema[BigDecimal] = bigdecimal
+    val foo = BigDecimal(1)
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+
+  }
+
+  test("smithy4s ByteArray") {
+    val schema: Schema[ByteArray] = bytes
+    val fooBar = ByteArray("fooBar".getBytes)
+    val hashOutput = visitor(schema).hash(fooBar)
+    expect.eql(fooBar.array.hashCode(), hashOutput)
+  }
+
+  test("smithy4s timestamp") {
+    val schema: Schema[Timestamp] = timestamp
+    val now = java.time.Instant.now()
+    val foo = Timestamp.fromEpochSecond(now.getEpochSecond)
+    val hashOutput = visitor(schema).hash(foo)
+    expect.eql(foo.hashCode(), hashOutput)
+  }
+
+
+
+  test("list") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+        struct(foos)(Foo.apply)
+      }.withId(ShapeId("", "Foo"))
+    }
+    val foo = Foo(List(1, 2, 3))
+    val hashOutput = visitor(Foo.schema).hash(foo)
+    expect.eql(foo.foos.hashCode(), hashOutput)
```

Comment:
This doesn't seem correct. A more correct expectation would be something along the lines of  `expect.eql(foo.hashCode, hashOutput)`

Reply:
I think the test cases should be changed to be just a list,set,vector etc.. and remove the struct

Reply:
the issue with mimicking scala case class hashcode method is that it uses the Product trait including the productPrefix method.
I can upcast all structs to Product and still use the forField method for each part
Currently I am using the productSeed from scala for combining hashes.

Reply:
> I can upcast all structs to Product 

You cannot guarantee that all structs are Scala products. Generally speaking, you must not make any assumptions on the in-memory representation during the implementation of Schema Visitors. 

It's fine that we cannot reproduce the native hashcode. HOWEVER, it's not fine that the hashcode of a product with one field is the same field as the hashcode of the value of the field. IE : `Foo(a).hash == a.hash` should be false, and your tests show that it's true.

Reply:
please forget about the existing tests. thats not my focus 
as is listed in the comment of this draft PR under todo . structs is included.
I am focused on what should the hash method look like.
 Are we trying to follow/mimic anything that currently exists i.e. Kittens implementation  of HList .
Or are we rolling our own with combining all hashes (in the appropriate fashion) including field labels and ShapeId (or any other alternative you might have)

Reply:
> as is listed in the comment of this draft PR under todo . structs is included.
 
Honestly it wouldn't hurt if you put a little care in how your PR descriptions look like. Github's markdown can help.

 > I am focused on what should the hash method look like.

Well all we need to do to get inspiration that is decompile the bytecode to find out how Scala implements it. A quick search gives me a SO thread with a promising, relatively recent [answer](https://stackoverflow.com/a/45311240/3065964). 

So the idea is probably to use the MurmurHash3 API to replicate a similar algorithm : 

1. start with an accumulator initialised to a static value (using `productSeed` is fine) 
2. iterate on the fields, mutating the accumulator by calling `mix` 
3. call the finalisation method to "avalanche" the bits of the hash (that seems to be the correct terminology). 

It's unlikely we could get our implementation that would fully respect the "must produce the same value as the native hashcode" invariant, because : 
* the official `Static` construct is annotated with a comment "not for public use", and I'd rather respect that. 
* our in-memory representation is not to be considered 
* if you take into consideration bijections/refinements, that invariant would be super easy to violate. 

However, this invariant should be respected for a subset of the metamodel, namely primitives and collections of primitives. This should be encoded in the tests. For the rest of the metamodel, it's gonna be reasonably hard to describe invariants that should be respected. At the very least we should test that : 

1. Wrapping a value in another layer of data (be that structure/union/collection) leads to the wrapped value having a different hashcode to the unwrapped value. 
2. Hashcodes of the different members of a union should be different. 
3. Computing the hashcode of a recursive data type doesn't blow the stack.

Reply:
> Honestly it wouldn't hurt if you put a little care in how your PR descriptions look like. Github's markdown can help.

Agreed